pkg/storage/edv/formatprovider/formatprovider.go

```diff
@@ -7,81 +7,122 @@ SPDX-License-Identifier: Apache-2.0
-	failCloseAllUnderlyingStores   = "failed to close all underlying stores in FormatProvider: %w"
-	failEncryptStructuredDocument  = "failed to encrypt structured document into a encrypted document: %w"
-	failMarshalEncryptedDocument   = "failed to marshal encrypted document into bytes: %w"
-	failPutInUnderlyingStore       = "failed to put encrypted document in underlying store in formatStore: %w"
-	failGetFromUnderlyingStore     = "failed to get encrypted document bytes from underlying store in formatStore: %w"
+	failOpenUnderlyingStore       = "failed to open underlying store in FormatProvider: %w"
+	failCloseUnderlyingStore      = "failed to close underlying store in FormatProvider: %w"
+	failCloseAllUnderlyingStores  = "failed to close all underlying stores in FormatProvider: %w"
+	failGenerateEDVCompatibleID   = "failed to generate EDV compatible ID: %w"
+	failEncryptStructuredDocument = "failed to encrypt structured document into a encrypted document: %w"
+	failMarshalEncryptedDocument  = "failed to marshal encrypted document into bytes: %w"
+	failPutInUnderlyingStore      = "failed to put encrypted document in underlying store in formatStore: %w"
+
 	failUnmarshalEncryptedDocument = "failed to unmarshal encrypted document bytes into encrypted document struct: %w"
 	failDecryptEncryptedDocument   = "failed to decrypt encrypted document into a structured document: %w"
 	failDeleteInUnderlyingStore    = "failed to delete key-value pair in underlying store in formatStore: %w"
 	failQueryUnderlyingStore       = "failed to query underlying store in formatStore: %w"
-	payloadKey                     = "payload"
+
+	payloadKey  = "payload"
+	keyIndexKey = "indexKey"
 )
 
 var (
 	errPayloadKeyMissing = errors.New(`the structured document content did not contain the ` +
 		`expected "payload" key`)
-	errPayloadNotAssertableAsByteArray = errors.New("unable to assert the payload value as a []byte")
+	errPayloadNotAssertableAsString = errors.New("unable to assert the payload value as a string")
 )
 
 type marshalFunc func(interface{}) ([]byte, error)
 
+// MACDigester represents a type that can compute MACs.
+type MACDigester interface {
+	ComputeMAC(data []byte, kh interface{}) ([]byte, error)
+}
+
+// MACCrypto is used for computing MACs.
+type MACCrypto struct {
+	kh          interface{}
+	macDigester MACDigester
+}
+
+// ComputeMAC computes a MAC for data using a matching MAC primitive in kh.
+func (m *MACCrypto) ComputeMAC(data string) (string, error) {
+	dataMAC, err := m.macDigester.ComputeMAC([]byte(data), m.kh)
+	return string(dataMAC), err
+}
+
 // DocumentProcessor represents a type that can encrypt and decrypt between
 // Structured Documents and Encrypted Documents.
 type DocumentProcessor interface {
-	Encrypt(*edv.StructuredDocument) (*edv.EncryptedDocument, error)
+	Encrypt(*edv.StructuredDocument, []edv.IndexedAttributeCollection) (*edv.EncryptedDocument, error)
```

Comment:
Just noticed these issues:

* DocumentProcessor is inside the `edv/formatprovider` package instead of being a first-class `storage` citizen (like we discussed previously). To recall: the "document processor" encrypts all artifacts into a standard form and would be reused with other storage implementations
* It looks like the documentprocessor does not encrypt the indexed attribute collection. Does it really need to be an argument? It's not even returned from `Decrypt()`, so there is an asymmetry here

Reply:
- The DocumentProcessor is used by the EDVFormatProvider, which allows the EDV format to be reused with any other storage implementation. There's no need for a "standard form".
- The indexed attributes are hashed before being passed in to the DocumentProcessor. This is what allows the documents to be queried. This allows us to still get documents based on a key per the usual provider method signature, so from the outside, an EDVFormatProvider (with whatever underlying provider you want) still works the same as any other provider. StructuredDocuments don't have encrypted indices, so they aren't returned.

Reply:
Now, I could create some sort of even more general DocumentProcessor class that just assumes JWE as what's being used for encryption. Or I could make it even more general and have it deal with just []byte. What do you think?

Reply:
Just realized one potentially confusing thing: the EDV spec calls them "Encrypted Indices", although they're actually HMACed.

Reply:
Note that the EDVFormatProvider will still require the more specific DocumentProcessor that deals with Structured and Encrypted Documents per the spec

Reply:
@DRK3 

> Now, I could create some sort of even more general DocumentProcessor class that just assumes JWE as what's being used for encryption. Or I could make it even more general and have it deal with just []byte. What do you think?

Last time we agreed we'd use the EDV format across all stores. The format would be pluggable, thus we'd have an "edv format" and potentially other formats. All would adhere to some common interface. This interface would a first-class citizen under `storage` or somewhere not far from it.

I guess we can regroup on monday.

Reply:
Output of monday's session will be a GitHub issue with the high-level tasks and design

Reply:
Wrote down general direction here: https://github.com/hyperledger/aries-framework-go/issues/2199#issuecomment-716617512

Reply:
Created an issue and added a ToDo above the FormatProvider: https://github.com/hyperledger/aries-framework-go/issues/2273

Reply:
My next PR will be to generalize this FormatProvider. This will involve moving it out of the `edv` folder and moving logic around to different places